### PR TITLE
Dynamic heap for ESP Wi-Fi

### PR DESCRIPTION
**Pull request by marekmatej**
The Zephyr's system heap is allocated statically and it is not possible to add the memory areas which are left unoccupied after all static allocations and bootstrapping.

This PR adds the possibility to re-claim those areas dynamically.
- Added Kconfig `CONFIG_ESP_RUNTIME_HEAP` (enabled by default)
- Utilize memory left after all static allocations and use it as Wi-Fi heap.
- Added choice Kconfig `CONFIG_ESP_WIFI_HEAP_*` to select the heap that WiFi adapter is using. Value `CONFIG_ESP_WIFI_HEAP_RUNTIME` is the default option.

All the additional system heap increments can be freed and used for other purposes. The impact on DRAM usage for the most commonly used ESP32 boards is summarized here. (The `samples/net/wifi` was used).

#### `ESP32-S3_DevkitC/esp32/procpu`
**Without** the dynamic ESP heap:
```
Memory region         Used Size  Region Size  %age Used
     iram0_0_seg:       46796 B       329 KB     13.89%
     dram0_0_seg:      163368 B       313 KB     50.97%
```

**With** the dynamic ESP heap:
```
Memory region         Used Size  Region Size  %age Used
     iram0_0_seg:       46852 B       329 KB     13.91%
     dram0_0_seg:      116336 B       313 KB     36.30%
```

#### `ESP32_Devkitc_wroom/esp32/procpu`

**Without** the dynamic ESP heap:
```
Memory region         Used Size  Region Size  %age Used
     iram0_0_seg:       64768 B       235 KB     26.91%
     dram0_0_seg:      124128 B       192 KB     63.13%
```

**With** the dynamic ESP heap:
```
Memory region         Used Size  Region Size  %age Used
     iram0_0_seg:       65024 B       224 KB     28.35%
     dram0_0_seg:       81408 B       192 KB     41.41%
```

#### `ESP32C3_DevkitC`
     
**without** the dynamic ESP heap:
```
Memory region         Used Size  Region Size  %age Used
     iram0_0_seg:      117264 B     301568 B     38.88%
     dram0_0_seg:      240320 B     301568 B     79.69%
```
     
**With** the dynamic ESP heap:
```
Memory region         Used Size  Region Size  %age Used
     iram0_0_seg:      117328 B     301568 B     38.91%
     dram0_0_seg:      197392 B     301568 B     65.46%
```

### CODE REVIEW

File: soc/espressif/esp32s3/default.ld [hunk 1]
```diff
@@ -952,8 +945,8 @@ SECTIONS
 ASSERT(((_iram_end - ORIGIN(iram0_0_seg)) <= LENGTH(iram0_0_seg)),
           "IRAM0 segment data does not fit.")
 
-ASSERT(((_heap_start - ORIGIN(dram0_0_seg)) <= LENGTH(dram0_0_seg)),
-          "DRAM segment data does not fit.")
+//ASSERT(((_heap_start - ORIGIN(dram0_0_seg)) <= LENGTH(dram0_0_seg)),
```
Comment (sylvioalves) [inline soc/espressif/esp32s3/default.ld:948]: Should this be removed?

**Comment by sylvioalves** [comment]
Overall LGTM. I have a question though: What if `CONFIG_ESP32_WIFI_HEAP_MEM_POOL=n`? Of course I understand this situation would not be recommended but it is an option. I think this would block Wi-Fi from working properly. Can you confirm?

**Comment by marekmatej** [comment]
> Overall LGTM. I have a question though: What if `CONFIG_ESP32_WIFI_HEAP_MEM_POOL=n`? Of course I understand this situation would not be recommended but it is an option. I think this would block Wi-Fi from working properly. Can you confirm?

In case `CONFIG_ESP32_WIFI_HEAP_MEM_POOL=n`, the default system heap will be used, and it is up to the system developer to have enough resources to accommodate WiFi.

However, in subsequent PR I would like to provide more options for WiFi heap, so it won't be limited by the dynamic heap size or usage.

**Review by wmrsouza — CHANGES_REQUESTED** [review]
Please, include ESP32-S3-DevKitC

**Comment by marekmatej** [comment]
> Please, include ESP32-S3-DevKitC

Thx, I missed that board. Updated! Please revisit.

**Review by wmrsouza — DISMISSED** [review]

**Review by uLipe — DISMISSED** [review]

**Review by raffarost — DISMISSED** [review]

**Review by raffarost — DISMISSED** [review]

**Review by uLipe — DISMISSED** [review]

**Review by raffarost — DISMISSED** [review]

File: soc/espressif/common/Kconfig [hunk 2]
```diff
@@ -24,6 +24,15 @@ config ESP_SIMPLE_BOOT
...
 
+config ESP_RUNTIME_HEAP
+	bool
+	default y
+	help
+	  Enabling this will allocate SRAM area starting by a last linked data at symbolic `_end`,
+	  ending by a last memory location that can be safely accesed (depending on a boot mode).
+	  This is a memory pool used in runtime to create a new heap memory.
+
+
```
Comment (sylvioalves) [inline soc/espressif/common/Kconfig:35]: Extra line here

**Review by sylvioalves — DISMISSED** [review]

**Review by raffarost — APPROVED** [review]

**Review by sylvioalves — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**